src/go/k8s/controllers/redpanda/cluster_controller.go

```diff
@@ -333,6 +341,140 @@ func (r *ClusterReconciler) setLicense(
+	ctx context.Context, rp *redpandav1alpha1.Cluster, log logr.Logger,
+) error {
+	pods, err := r.podList(ctx, rp)
+	if err != nil {
+		return fmt.Errorf("unable to fetch PodList: %w", err)
+	}
+	for i := range pods.Items {
+		pod := &pods.Items[i]
+		if pod.DeletionTimestamp.IsZero() {
+			// if the pod is not being deleted, set the finalizer
+			if err = r.setPodFinalizer(ctx, pod); err != nil {
+				//nolint:goerr113 // not going to use wrapped static error here this time
+				return fmt.Errorf(`unable to set the finalizer on pod "%s": %d`, pod.Name, err)
+			}
+			continue
+		}
+		// if the pod is being deleted
+		// check the node it's assigned to
+		node := corev1.Node{}
+		key := types.NamespacedName{Name: pod.Spec.NodeName}
+		err := r.Get(ctx, key, &node)
+		// if the node is not gone
+		if err != nil && !apierrors.IsNotFound(err) {
+			return fmt.Errorf(`unable to fetch node "%s": %w`, pod.Spec.NodeName, err)
+		}
+		// nor has a noexecute taint
+		untainted := true
+		for _, taint := range node.Spec.Taints {
+			if taint.Effect == corev1.TaintEffectNoExecute && taint.Key == corev1.TaintNodeUnreachable {
+				untainted = false
+			}
+		}
+		if untainted {
+			// remove the finalizer and let the pod be restarted
+			if err = r.removePodFinalizer(ctx, pod); err != nil {
+				return fmt.Errorf(`unable to remove finalizer from pod "%s": %w`, pod.Name, err)
+			}
+			continue
+		}
+		// get the node id
+		nodeIDStr, ok := pod.GetAnnotations()[PodAnnotationNodeIDKey]
+		if !ok {
+			return fmt.Errorf("cannot determine node_id for pod %s: %w. not removing finalizer", pod.Name, err)
+		}
+		nodeID, err := strconv.Atoi(nodeIDStr)
+		if err != nil {
+			return fmt.Errorf("node-id annotation is not an integer: %w", err)
+		}
+		// get the broker list
+		redpandaPorts := networking.NewRedpandaPorts(rp)
+		headlessPorts := collectHeadlessPorts(redpandaPorts)
+		clusterPorts := collectClusterPorts(redpandaPorts, rp)
+		headlessSvc := resources.NewHeadlessService(r.Client, rp, r.Scheme, headlessPorts, log)
+		clusterSvc := resources.NewClusterService(r.Client, rp, r.Scheme, clusterPorts, log)
+
+		pki := certmanager.NewPki(r.Client, rp, headlessSvc.HeadlessServiceFQDN(r.clusterDomain), clusterSvc.ServiceFQDN(r.clusterDomain), r.Scheme, log)
+
+		adminClient, err := r.AdminAPIClientFactory(ctx, r.Client, rp, headlessSvc.HeadlessServiceFQDN(r.clusterDomain), pki.AdminAPIConfigProvider())
+		if err != nil {
+			return fmt.Errorf("unable to create admin client: %w", err)
+		}
+		brokers, err := adminClient.Brokers(ctx)
+		if err != nil {
+			return fmt.Errorf("unable to fetch brokers: %w", err)
+		}
+		// check if the node in the broker list
+		var broker *admin.Broker
+		for i := range brokers {
+			if brokers[i].NodeID == nodeID {
+				broker = &brokers[i]
+				break
+			}
+		}
+		// if it's not gone
+		if broker != nil {
+			// decommission it
+			if err = adminClient.DecommissionBroker(ctx, nodeID); err != nil {
```

Comment:
If we are decommissioning broker will the maintenance mode from POD preStop script prevent or influence the decommission process?

Reply:
BTW decommission might never finish if there is 3 node cluster and partition with replication of 3. In such case decommission is stuck. 

This API is asynchronous. Once you issue decommission you need to watch out for brokers (`adminClient.Brokers(ctx)`) to finish decommission process. I'm not sure if you can call this API multiple times. Definitely we can not revert this operation once it is started.

Reply:
Right now it will not matter. There was already logic to come through and reset the maintenance mode for decommissioned brokers.